// ngx-localstorage/src/ngx-localstorage.module.js
/**
 * Created by bohoffi on 03.04.2017.
 */
import { InjectionToken, NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LocalStorageDirective, LocalStorageService } from './lib/index';
export var ModuleConfigToken = new InjectionToken('moduleConfig');
export function provideStorageService(moduleConfig) {
    return new LocalStorageService(moduleConfig);
}
var NgxLocalStorageModule = (function () {
    function NgxLocalStorageModule(parentModule) {
        if (parentModule) {
            throw new Error('NgxLocalStorageModule is already loaded. Import it in the AppModule only');
        }
    }
    NgxLocalStorageModule.forRoot = function (config) {
        return {
            ngModule: NgxLocalStorageModule,
            providers: [
                {
                    provide: ModuleConfigToken,
                    useValue: config
                },
                {
                    provide: LocalStorageService,
                    useFactory: provideStorageService,
                    deps: [
                        ModuleConfigToken
                    ]
                }
            ]
        };
    };
    return NgxLocalStorageModule;
}());
export { NgxLocalStorageModule };
NgxLocalStorageModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    LocalStorageDirective
                ],
                exports: [
                    LocalStorageDirective
                ]
            },] },
];
/** @nocollapse */
NgxLocalStorageModule.ctorParameters = function () { return [
    { type: NgxLocalStorageModule, decorators: [{ type: Optional }, { type: SkipSelf },] },
]; };
export { LocalStorageService } from './lib/index';
export { ngxLocalStorage } from './lib/decorators';
